src/qt/splashscreen.cpp

```diff
@@ -170,27 +170,28 @@ static void InitMessage(SplashScreen *splash, const std::string &message)
         Q_ARG(QColor, QColor(55,55,55)));
 }
 
-static void ShowProgress(SplashScreen *splash, const std::string &title, int nProgress)
-{
-    InitMessage(splash, title + strprintf("%d", nProgress) + "%");
-}
-
 void SplashScreen::setBreakAction(const std::function<void(void)> &action)
 {
     breakAction = action;
 }
 
-static void SetProgressBreakAction(SplashScreen *splash, const std::function<void(void)> &action)
+static void ShowProgress(SplashScreen *splash, const std::string &title, int nProgress, bool cancel_possible)
 {
-    QMetaObject::invokeMethod(splash, "setBreakAction",
-        Qt::QueuedConnection,
-        Q_ARG(std::function<void(void)>, action));
+    InitMessage(splash, title + strprintf("%d", nProgress) + "%" +
+            (cancel_possible ? (std::string("\n") + _("(press q to shutdown and continue later)"))
+                                : std::string("")));
+
+    if (cancel_possible) {
```

Comment:
Don't you need to invoke `setBreakAction(nullptr)` here when `cancel_possible` is false?

Maybe better would be to get rid of this `std::function` madness and replace the `SplashScreen::breakAction` pointer with a simpler boolean `std::atomic<bool> m_cancel_possible`.